src/services/earn/earn-service.ts

```diff
@@ -613,28 +618,47 @@ export class EarnService implements IEarnService {
       calls.push(buildPermissionPermit(bigIntPositionId, permissionPermit, vault));
     }
 
-    // Handle reduce
+    // Handle withdraw
     calls.push(
       encodeFunctionData({
         abi: companionAbi,
         functionName: 'withdraw',
-        args: [vault as ViemAddress, bigIntPositionId, tokensToWithdraw, intendedWithdraw, COMPANION_SWAPPER_CONTRACT.address(chainId)],
+        args: [
+          vault as ViemAddress,
+          bigIntPositionId,
+          tokensToWithdraw,
+          intendedWithdraw.map((amount, index) => (withdraw.amounts[index].type != WithdrawType.MARKET ? amount : 0n)),
+          COMPANION_SWAPPER_CONTRACT.address(chainId),
+        ],
       })
     );
 
-    let balancesFromVault: Record<TokenAddress, bigint> = {};
-    // If any token amount to convert is MAX_UINT256, then we need to check vault balance
-    if (withdraw.amounts.some(({ convertTo, amount }) => !!convertTo && amount == Uint.MAX_256)) {
-      // position = [ tokens, balances, strategy ]
-      const position = await this.providerService.getViemPublicClient({ chainId }).readContract({
-        abi: vaultAbi,
-        address: vault,
-        functionName: 'position',
-        args: [bigIntPositionId],
-      });
-      balancesFromVault = Object.fromEntries(position[0].map((token, index) => [token, position[1][index]]));
+    // Handle special withdraw
+    if (shouldWithdrawFarmToken) {
+      calls.push(
+        encodeFunctionData({
+          abi: companionAbi,
+          functionName: 'specialWithdraw',
+          args: [
+            vault as ViemAddress,
+            bigIntPositionId,
+            BigInt(SpecialWithdrawalCode.WITHDRAW_ASSET_FARM_TOKEN_BY_AMOUNT),
```

Comment:
I think we should be using `WITHDRAW_ASSET_FARM_TOKEN_BY_ASSET_AMOUNT`. The FE doesn't know how many farm tokens we have, right?
```suggestion
            BigInt(SpecialWithdrawalCode.WITHDRAW_ASSET_FARM_TOKEN_BY_ASSET_AMOUNT),
```